src/AddIns/Uno.WinUI.Graphics3DGL/GLCanvasElement.cs

```diff
@@ -220,8 +221,27 @@ private void OnClosed(object _, object __)
 	/// </summary>
 #if WINAPPSDK
 	public void Invalidate() => DispatcherQueue.TryEnqueue(DispatcherQueuePriority.Low, Render);
-#else // WPF hangs if we attempt to enqueue on Low inside RenderOverride
-	public void Invalidate() => NativeDispatcher.Main.Enqueue(Render, NativeDispatcherPriority.Idle);
+#else
+	// We used to Invalidate like this
+	// public void Invalidate() => NativeDispatcher.Main.Enqueue(Render, NativeDispatcherPriority.Idle);
+	// but Win32 will hang when resizing or moving the window by dragging it
+	// because the DefWindowProc handler for WM_SYSCOMMAND will enter an internal loop that processes received
+	// win32 messages until the pointer that resizes/moves the window is released. In that loop, our
+	// UnoWin32DispatcherMsg messages are always processed before WM_PAINT messages, so if user code calls
+	// Invalidate() inside RenderOverride(), we will enter an infinite loop that completely hangs the app.
+	public void Invalidate()
+	{
+		Compositor.GetSharedCompositor().InvalidateRender(Visual);
```

Comment:
in this case, we should not invalidate if there's already compositiontarget is currently animating, as it will automatically invalidate because the refresh timer.

Reply:
It won't. The call to InvalidateRender will also call InvalidatePaint on the visual. Otherwise, the visual won't be repainted.